cirq-core/cirq/transformers/merge_single_qubit_gates_test.py

```diff
@@ -0,0 +1,139 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List
+
+import numpy as np
+import pytest
+
+import cirq
+
+
+def assert_optimizes(optimized: cirq.AbstractCircuit, expected: cirq.AbstractCircuit):
+    # Ignore differences that would be caught by follow-up optimizations.
+    followup_transformers: List[cirq.TRANSFORMER] = [
+        cirq.drop_negligible_operations,
+        cirq.drop_empty_moments,
+    ]
+    for transform in followup_transformers:
+        optimized = transform(optimized)
+        expected = transform(expected)
+
+    cirq.testing.assert_same_circuits(optimized, expected)
+
+
+def test_not_both():
+    with pytest.raises(ValueError, match='both rewriter and synthesizer'):
+        _ = cirq.merge_single_qubit_gates(
+            cirq.Circuit(), synthesizer=lambda *args: None, rewriter=lambda *args: None
+        )
+
+
+def test_combines_sequence():
+    q = cirq.NamedQubit('q')
+    c = cirq.Circuit(cirq.X(q) ** 0.5, cirq.Z(q) ** 0.5, cirq.X(q) ** -0.5)
+    c = cirq.merge_single_qubit_gates(c)
+    op_list = [*c.all_operations()]
+    assert len(op_list) == 1
+    assert isinstance(op_list[0].gate, cirq.MatrixGate)
+    cirq.testing.assert_allclose_up_to_global_phase(
+        cirq.unitary(c), cirq.unitary(cirq.Y ** 0.5), atol=1e-7
+    )
+
+
+def test_removes_identity_sequence():
+    q = cirq.NamedQubit('q')
+    assert_optimizes(
+        optimized=cirq.merge_single_qubit_gates(
+            cirq.Circuit([cirq.Z(q), cirq.H(q), cirq.X(q), cirq.H(q)])
+        ),
+        expected=cirq.Circuit(),
+    )
+
+
+def test_stopped_at_2qubit():
+    q, q2 = cirq.LineQubit.range(2)
+    c = cirq.Circuit([cirq.Z(q), cirq.H(q), cirq.X(q), cirq.H(q), cirq.CZ(q, q2), cirq.H(q)])
+    c = cirq.drop_empty_moments(cirq.merge_single_qubit_gates(c))
+    assert len(c) == 3
+    cirq.testing.assert_allclose_up_to_global_phase(cirq.unitary(c[0]), np.eye(2), atol=1e-7)
+    assert isinstance(c[-1][q].gate, cirq.MatrixGate)
+
+
+def test_respects_nocompile_tags():
+    q = cirq.NamedQubit("q")
+    c = cirq.Circuit(
+        [cirq.Z(q), cirq.H(q), cirq.X(q), cirq.H(q), cirq.X(q).with_tags("nocompile"), cirq.H(q)]
+    )
+    context = cirq.TransformerContext(tags_to_ignore=("nocompile",))
+    c = cirq.drop_empty_moments(cirq.merge_single_qubit_gates(c, context=context))
+    assert len(c) == 3
+    cirq.testing.assert_allclose_up_to_global_phase(cirq.unitary(c[0]), np.eye(2), atol=1e-7)
+    assert c[1][q] == cirq.X(q).with_tags("nocompile")
+    assert isinstance(c[-1][q].gate, cirq.MatrixGate)
+
+
+def test_ignores_2qubit_target():
+    c = cirq.Circuit(cirq.CZ(*cirq.LineQubit.range(2)))
+    assert_optimizes(optimized=cirq.merge_single_qubit_gates(c), expected=c)
+
+
```

Comment:
Could we add a few tests for each optimizer to ensure they properly ignore measurements ?

Reply:
done